qiskit_nature/algorithms/initial_points/mp2_initial_point.py

```diff
@@ -0,0 +1,302 @@
+
+import numpy as np
+from qiskit_nature.exceptions import QiskitNatureError
+
+from qiskit_nature.properties.second_quantization.electronic import ElectronicEnergy
+from qiskit_nature.properties.second_quantization.electronic.bases import ElectronicBasis
+from qiskit_nature.circuit.library import UCC
+from qiskit_nature.properties.second_quantization.second_quantized_property import (
+    GroupedSecondQuantizedProperty,
+)
+
+from .initial_point import InitialPoint
+
+
+class MP2InitialPoint(InitialPoint):
+    """Moller-Plesset 2nd order initial point generator.
+
+    A class for computing the Moller-Plesset second order (MP2) corrections to use as an initial
+    point with a VQE algorithm alongside a UCC ansatz.
+
+    The MP2 calculation requires the two-body molecular orbital matrix, the orbital energies,
+    and the Hartree-Fock reference energy from the ``ElectronicEnergy`` in the grouped
+    property.
+
+    An MP2InitialPoint uses the same excitations as generated by the UCC ansatz to ensure that
+    the coefficients are mapped correctly in the initial point. Only initial point values that
+    correspond to double-excitations will be non-zero.
+
+    The coefficients and energy deltas are computed using the ``get_initial_point`` method.
+    Thereafter, they can be accessed via the ``initial_point`` and ``energy_deltas`` properties.
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.threshold = 1e-12
+        self._corrections: dict[str, tuple[float, float]] = None
+
+    @property
+    def grouped_property(self) -> GroupedSecondQuantizedProperty:
+        """The grouped property."""
+        return self._grouped_property
+
+    @grouped_property.setter
+    def grouped_property(self, grouped_property: GroupedSecondQuantizedProperty) -> None:
+        """The grouped property.
+
+        The grouped property is required to contain the ElectronicEnergy, which must contain
+        the two-body molecular orbital matrix, the orbital energies, and the Hartree-Fock
+        reference energy.
+
+        Raises:
+            QiskitNatureError: If the ``ElectronicEnergy`` is missing or the two-body molecular
+            orbital matrix, the orbital energies, and/or the Hartree-Fock reference energy are
+            missing from ``ElectronicEnergy``.
+        """
+        try:
+            electronic_energy: ElectronicEnergy = grouped_property.get_property(ElectronicEnergy)
+        except TypeError as invalid_grouped_property:
+            raise QiskitNatureError from invalid_grouped_property
+
+        try:
+            self._integral_matrix: np.ndarray = electronic_energy.get_electronic_integral(
+                ElectronicBasis.MO, 2
+            ).get_matrix()
+            # Infer the number of molecular orbitals from the MO matrix.
+            self._num_molecular_orbitals: int = self._integral_matrix.shape[0]
+        except TypeError as missing_two_body_mo_matrix:
```

Comment:
Here, the error which will be raised is actually an `AttributeError` because the missing integral will again be `None` resulting in `get_matrix()` being called on a `NoneType`.

I suggest you also add some unittests which specifically trigger these exceptions

Reply:
I've overhauled these checks and added unit tests specific to each error raised.